comp110-worksheet-6/OxoBoard.cs

```diff
@@ -10,49 +10,139 @@ public enum Mark { None, O, X };
+                    board[x,y] = mark;
+                    return true;
+                }
+            }
+
+            return false;
 		}
 
 		// If there are still empty squares on the board, return false.
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            for (int x = 0; x < board.GetLength(0); x++)
+            {
+                for (int y = 0; y < board.GetLength(1); y++)
+                {
+                    if (board[x,y] == Mark.None)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
 		}
 
 		// If a player has three in a row, return Mark.O or Mark.X depending on which player.
 		// Otherwise, return Mark.None.
 		public Mark GetWinner()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+            foreach (Mark mark in new Mark[] { Mark.O, Mark.X })
+            {
+                if (CheckNegDiagonalMark(mark) || CheckPosDiagonalMark(mark))
+                {
+                    return mark;
+                }
+
+                for (int i = 0; i < board.GetLength(0); i++)
+                {
+                    if (CheckHorizontalMark(mark, i) || CheckVerticalMark(mark, i))
+                    {
+                        return mark;
+                    }
+                }
+            }
+
+            return Mark.None;
 		}
 
-		// Display the current board state in the terminal. You should only need to edit this if you are attempting the stretch goal.
-		public void PrintBoard()
+        public bool CheckHorizontalMark(Mark mark, int y)
+        {
+            for (int x = 0; x < board.GetLength(0); x++)
+            {
+                if (board[x, y] != mark)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public bool CheckVerticalMark(Mark mark, int x)
+        {
+            for (int y = 0; y < board.GetLength(1); y++)
+            {
+                if (board[x, y] != mark)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public bool CheckNegDiagonalMark(Mark mark)
+        {
+            for (int i = 0; i < board.GetLength(0); i++)
+            {
+                if (board[i, i] != mark)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public bool CheckPosDiagonalMark(Mark mark)
+        {
+            for (int i = 0; i < board.GetLength(0); i++)
+            {
+                if (board[board.GetLength(0) - 1 - i, i] != mark)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
```

Comment:
These functions are a little repetitive -- could they be refactored to share code?